ocaml/xenopsd/lib/xenops_task.ml

```diff
@@ -39,28 +39,27 @@ let event_wait local_updates task ?from ?timeout_start timeout p =
           (Some (remaining |> ceil |> int_of_float))
           local_updates
       in
-      let success = List.fold_left (fun acc d -> acc || p d) false deltas in
-      let finished = success in
-      if not finished then
-        let elapsed = Unix.gettimeofday () -. start in
-        inner (timeout -. elapsed) (Some next_id)
-      else
-        success
+      match List.find_map filter deltas with
+      | Some success when p success ->
+          Some success
```

Comment:
We get here if we have detected a change in state on the task, and the task has ended (either succeeded or failed) within the specified timeout.